src/vm/types/mod.rs

```diff
@@ -0,0 +1,450 @@
+    pub fn type_result(&self) -> TypeSignature {
+        use self::AtomTypeIdentifier::*;
+        use self::BlockInfoProperty::*;
+        TypeSignature::from(
+            match self {
+                Time => IntType,
+                VrfSeed => BufferType(32),
+                HeaderHash => BufferType(32),
+                BurnchainHeaderHash => BufferType(32),
+            })
+    }
+}
+
+impl PartialEq for ListData {
+    fn eq(&self, other: &ListData) -> bool {
+        self.data == other.data
+    }
+}
+
+impl Hash for ListData {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.data.hash(state);
+    }
+}
+
+impl PartialEq for TupleData {
+    fn eq(&self, other: &TupleData) -> bool {
+        self.data_map == other.data_map
+    }
+}
+
+impl Hash for TupleData {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.data_map.hash(state);
+    }
+}
+
+pub const NONE: Value = Value::Optional(OptionalData { data: None });
+
+impl Value {
+    pub fn some(data: Value) -> Value {
+        Value::Optional(OptionalData {
+            data: Some(Box::new(data)) })
+    }
+
+    pub fn none() -> Value {
+        Value::Optional(OptionalData {
+            data: None })
+    }
+
+    pub fn okay(data: Value) -> Value {
+        Value::Response(ResponseData { 
+            committed: true,
+            data: Box::new(data) })
+    }
+
+    pub fn error(data: Value) -> Value {
+        Value::Response(ResponseData { 
+            committed: false,
+            data: Box::new(data) })
+    }
+
+    /// Invariant: the supplied Values have already been "checked", i.e., it's a valid Value object
+    ///  this invariant is enforced through the Value constructors, each of which checks to ensure
+    ///  that any typing data is correct.
+    pub fn list_with_type(list_data: Vec<Value>, expected_type: ListTypeData) -> Result<Value> {
+        if expected_type.size()? > MAX_VALUE_SIZE {
+            return Err(RuntimeErrorType::ValueTooLarge.into())
+        }
+
+        if (expected_type.max_len as usize) < list_data.len() {
+            return Err(InterpreterError::FailureConstructingListWithType.into())
+        }
+
+        let expected_item_type = expected_type.get_list_item_type();
+
+        for item in &list_data {
+            if !expected_item_type.admits(&item) {
+                return Err(InterpreterError::FailureConstructingListWithType.into())
+            }
+        }
+
+        Ok(Value::List(ListData { data: list_data, type_signature: expected_type }))
+    }
+
+    pub fn list_from(list_data: Vec<Value>) -> Result<Value> {
+        let type_sig = TypeSignature::construct_parent_list_type(&list_data)?;
+        // Aaron: at this point, we've _already_ allocated memory for this type.
+        //     (e.g., from a (map...) call, or a (list...) call.
+        //     this is a problem _if_ the static analyzer cannot already prevent
+        //     this case. This applies to all the constructor size checks.
```

Comment:
Would this mean that a high-dimensional list with one large dimension value could either (1) trigger OOM or (2) bill the contract creator with a high tx fee as if the list were a high-dimensional "cube", like before?  We probably want to avoid both (1) and (2) if at all possible.

Reply:
Yep -- the type checker will need to save us in this case, which it will be able to do (once we merge https://github.com/blockstack/blockstack-core/tree/feature/value-size-bounding), since two properties are true:

1. TypeSignatures will Error rather than construct a too-large-value.
2. Every node in the AST will be associated with a TypeSignature following a successful type_checker pass.